rasa_core/agent.py

```diff
@@ -419,6 +468,85 @@ def _create_ensemble(policies):
         else:
             passed_type = type(policies).__name__
             raise ValueError(
-                    "Invalid param `policies`. Passed object is "
-                    "of type '{}', but should be policy, an array of "
-                    "policies, or a policy ensemble".format(passed_type))
+                "Invalid param `policies`. Passed object is "
+                "of type '{}', but should be policy, an array of "
+                "policies, or a policy ensemble".format(passed_type))
+
+    @staticmethod
+    def _get_model_hash(model_server, header='model_hash'):
+        # type: (Text, Text) -> Optional[Text]
+        head = requests.head(model_server)
+        head.raise_for_status()
+        return head.headers.get(header)
+
+    @staticmethod
+    def init_model_from_server(model_server,  # type: Text
+                               action_factory,  # type: Text
+                               model_directory  # type: Text
+                               ):
+        # type: (...) -> Optional[(Domain, PolicyEnsemble, Text)]
+        """Initialises a Rasa Core model from a URL."""
+
+        if not is_url(model_server):
+            raise requests.exceptions.InvalidURL(model_server)
+        try:
+            new_hash = Agent._get_model_hash(model_server)
+            Agent._pull_model(model_server, model_directory)
+            domain_path = os.path.join(os.path.abspath(model_directory),
+                                       "domain.yml")
+            domain = TemplateDomain.load(domain_path, action_factory)
+            policy_ensemble = PolicyEnsemble.load(model_directory)
+            return domain, policy_ensemble, new_hash
+        except Exception as e:
+            logger.exception("Could not retrieve model from server "
+                             "URL {}:\n{}".format(model_server, e))
+            raise e
+
+    def _update_model_from_server(self,
+                                  model_server,  # type: Text
+                                  action_factory,  # type: Text
+                                  model_directory  # type: Text
+                                  ):
+        # type: (...) -> Optional[(Domain, PolicyEnsemble, Text)]
+        """Loads a zipped Rasa Core model from a URL."""
+
+        if not is_url(model_server):
+            raise requests.exceptions.InvalidURL(model_server)
+        new_hash = Agent._get_model_hash(model_server)
+        if new_hash != self.model_hash:
+            self.model_hash = new_hash
```

Comment:
isn't this duplicating the code from `init_model_from_server`

Reply:
apart from the first two lines the methods are quite different. `update_model` assumes state, `init_model` doesn't